File: src/services/audit/detectors/ad/advanced/signing/ldap-channel-binding-disabled.ts
/**
 * LDAP Channel Binding Disabled Detector
 * Check if LDAP channel binding is disabled
 * Requires GPO settings from SYSVOL
 */

import { ADDomain } from '../../../../../../types/ad.types';
import { Finding } from '../../../../../../types/finding.types';
import { GpoSecuritySettings } from '../types';

export function detectLdapChannelBindingDisabled(
  gpoSettings: GpoSecuritySettings | null,
  domain: ADDomain | null,
  includeDetails: boolean
): Finding {
  if (gpoSettings && gpoSettings.ldapChannelBinding !== undefined) {
    const bindingDisabled = gpoSettings.ldapChannelBinding === 0;

    return {
      type: 'LDAP_CHANNEL_BINDING_DISABLED',
      severity: 'high',
      category: 'advanced',
      title: 'LDAP Channel Binding Not Required',
      description:
        'LDAP channel binding is not required. This allows NTLM relay attacks against LDAPS connections even when signing is enabled.',
      count: bindingDisabled ? 1 : 0,
      affectedEntities: includeDetails && bindingDisabled && domain ? [domain.dn] : undefined,
      details: bindingDisabled
        ? {
            recommendation: 'Configure "Domain controller: LDAP server channel binding token requirements" to "Always".',
            currentSetting: gpoSettings.ldapChannelBinding,
            requiredSetting: 2,
          }
        : undefined,
    };
  }

  return {
    type: 'LDAP_CHANNEL_BINDING_DISABLED',
    severity: 'high',
    category: 'advanced',
    title: 'LDAP Channel Binding Configuration Unknown',
    description: 'Unable to determine LDAP channel binding configuration. Manual review recommended.',
    count: 0,
    details: {
      note: 'GPO/Registry settings not available. Check LdapEnforceChannelBinding registry value manually.',
    },
  };
}

File: internal/audit/detectors/ad/advanced/signing/ldap-channel-binding-disabled.go
package signing

import (
	"context"

	"github.com/etcsec-com/etc-collector/internal/audit"
	"github.com/etcsec-com/etc-collector/pkg/types"
)

// LdapChannelBindingDisabledDetector detects if LDAP channel binding is disabled
type LdapChannelBindingDisabledDetector struct {
	audit.BaseDetector
}

// NewLdapChannelBindingDisabledDetector creates a new detector
func NewLdapChannelBindingDisabledDetector() *LdapChannelBindingDisabledDetector {
	return &LdapChannelBindingDisabledDetector{
		BaseDetector: audit.NewBaseDetector("LDAP_CHANNEL_BINDING_DISABLED", audit.CategoryAdvanced),
	}
}

// Detect executes the detection
func (d *LdapChannelBindingDisabledDetector) Detect(ctx context.Context, data *audit.DetectorData) []types.Finding {
	// This requires GPO/registry settings - for now, flag for manual review
	finding := types.Finding{
		Type:        d.ID(),
		Severity:    types.SeverityHigh,
		Category:    string(d.Category()),
		Title:       "LDAP Channel Binding Configuration Unknown",
		Description: "Unable to determine LDAP channel binding configuration. Manual review recommended.",
		Count:       0,
		Details: map[string]interface{}{
			"note":           "GPO/Registry settings not available via LDAP. Check LdapEnforceChannelBinding registry value manually.",
			"recommendation": "Configure 'Domain controller: LDAP server channel binding token requirements' to 'Always'.",
			"registryPath":   "HKLM\\SYSTEM\\CurrentControlSet\\Services\\NTDS\\Parameters\\LdapEnforceChannelBinding",
			"requiredValue":  2,
		},
	}

	return []types.Finding{finding}
}

func init() {
	audit.MustRegister(NewLdapChannelBindingDisabledDetector())
}
